File: Java/opentraceability/src/main/java/opentraceability/mappers/OpenTraceabilityJsonLDMapper.java
package opentraceability.mappers;

import opentraceability.interfaces.*;
import opentraceability.models.events.kdes.*;
import opentraceability.models.events.*;
import opentraceability.models.identifiers.*;
import opentraceability.utility.*;
import opentraceability.models.common.*;
import opentraceability.*;
import org.apache.commons.codec.language.bm.Lang;
import org.json.JSONArray;
import org.json.JSONObject;
import tangible.StringHelper;


import java.time.Duration;
import java.time.OffsetDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

public final class OpenTraceabilityJsonLDMapper
{
	public static Object ToJson(Object value, Map<String, String> namespacesReversed, Boolean isGs1WebVocab, boolean required) throws Exception {
		if (value != null)
		{
			JSONObject json = new JSONObject();
			JSONObject jpointer = json;

			Class t = value.getClass();
			OTMappingTypeInformation typeInfo = OTMappingTypeInformation.getJsonTypeInfo(t);
			for (var property : typeInfo.properties.stream().filter(p -> p.version.equals(EPCISVersion.Any) || p.version == EPCISVersion.V2).collect(Collectors.toList()))
			{
				Object obj = property.field.get(value);
				if (obj != null)
				{
					JSONObject jvaluepointer = jpointer;
					String jChildName = property.name;

					if (property.isQuantityList)
					{
						ArrayList<EventProduct> products = (ArrayList<EventProduct>)obj;
						var quantityProducts = products.stream().filter(p -> p.Quantity != null && p.Type == property.productType).collect(Collectors.toList());
						if (!quantityProducts.isEmpty())
						{
							JSONArray jQuantityList = new JSONArray();
							for (var product : quantityProducts)
							{
								if (product.EPC != null && product.Quantity != null)
								{
									JSONObject jQuantity = new JSONObject();
									jQuantity.put("epcClass", product.EPC.toString());
									jQuantity.put("quantity", product.Quantity.value);
									if (!Objects.equals(product.Quantity.uom.UNCode, "EA"))
									{
										jQuantity.put("uom", product.Quantity.uom.UNCode);
									}
									jQuantityList.put(jQuantity);
								}
							}
							JSONExtensions.put(jvaluepointer, jChildName, jQuantityList);
						}
					}
					else if (property.isEPCList)
					{
						ArrayList<EventProduct> products = (ArrayList<EventProduct>)obj;
						var epcProducts = products.stream().filter(p -> p.Quantity == null && p.Type == property.productType).collect(Collectors.toList());
						if (!epcProducts.isEmpty() || property.required)
						{
							JSONArray jEPCList = new JSONArray();
							for (var product : epcProducts)
							{
								if (product.EPC != null)
								{
									JSONExtensions.put(jEPCList, product.EPC.toString());
								}
							}
							JSONExtensions.put(jvaluepointer, jChildName, jEPCList);
						}
					}
					else if (property.isArray)
					{
						List list = (List)obj;
						JSONArray jList = new JSONArray();

						if (!list.isEmpty() || property.required)
						{
							if (property.isRepeating && list.size() == 1)
							{
								Object jt = WriteObjectToJToken(list.get(0), isGs1WebVocab);
								if (jt != null)
								{
									JSONExtensions.put(jvaluepointer, jChildName, jt);
								}
							}
							else
							{
								for (var o : list)
								{
									if (property.isObject)
									{
										Object jListChild = ToJson(o, namespacesReversed, isGs1WebVocab, property.required);
										if (jListChild != null)
										{
											JSONExtensions.put(jList, jListChild);
										}
									}
									else if (property.itemType == LanguageString.class && isGs1WebVocab)
									{
										LanguageString ls = (LanguageString)o;
										JSONObject j = ls.toJSON();
										JSONExtensions.put(jList, j);
									}
									else
									{
										Object jt = WriteObjectToJToken(o, isGs1WebVocab);
										if (jt != null)
										{
											JSONExtensions.put(jList, jt);
										}
									}
								}

								JSONExtensions.put(jvaluepointer, jChildName, jList);
							}
						}
					}
					else if (property.isObject)
					{
						Object jChild = ToJson(obj, namespacesReversed, isGs1WebVocab, property.required);
						if (jChild != null)
						{
							JSONExtensions.put(jvaluepointer, jChildName, jChild);
						}
					}
					else
					{
						Object jt = WriteObjectToJToken(obj, isGs1WebVocab);
						if (jt != null)
						{
							JSONExtensions.put(jvaluepointer, jChildName, jt);
						}
					}
				}
			}

			if (typeInfo.extensionKDEs != null)
			{
				Object obj = typeInfo.extensionKDEs.get(value);
				if (obj != null)
				{
					ArrayList<IEventKDE> kdes = (ArrayList<IEventKDE>)obj;
					if (kdes != null)
					{
						for (var kde : kdes)
						{
							Object jChild = kde.getJson();
							if (jChild != null)
							{
								String name = kde.name;
								if (!StringHelper.isNullOrEmpty(kde.namespace))
								{
									if (!namespacesReversed.containsKey(kde.namespace))
									{
										throw new RuntimeException(String.format("The namespace %1$s is not recognized in the EPCIS Document / EPCIS Query Document.", kde.namespace));
									}
									name = namespacesReversed.get(kde.namespace) + ":" + name;
								}

								JSONExtensions.put(jpointer, name, jChild);
							}
						}
					}
				}
			}

			if (typeInfo.extensionAttributes != null)
			{
				Object obj = typeInfo.extensionAttributes.get(value);
				if (obj != null)
				{
					ArrayList<IEventKDE> kdes = (ArrayList<IEventKDE>)obj;
						if (kdes != null)
						{
							for (IEventKDE kde : kdes)
							{
								Object jChild = kde.getJson();
								if (jChild != null)
								{
									String name = kde.name;
									if (kde.namespace != null)
									{
										if (!namespacesReversed.containsKey(kde.namespace))
										{
											throw new RuntimeException(String.format("The namespace %1$s is not recognized in the EPCIS Document / EPCIS Query Document.", kde.namespace));
										}
										name = namespacesReversed.get(kde.namespace) + ":" + name;
									}

									JSONExtensions.put(jpointer, name, jChild);
								}
							}
						}
					}
				}

			return json;
		}
		else
		{
			return null;
		}
	}

	/** 
	 Converts a JSON object into the type specified.
	*/
	public static Object FromJson(Object json, Class type, Map<String, String> namespaces) throws Exception {
		Object value = ReflectionUtility.constructType(type);

		OTMappingTypeInformation typeInfo = OTMappingTypeInformation.getJsonTypeInfo(type);

		ArrayList<IEventKDE> extensionKDEs = null;
		ArrayList<IEventKDE> extensionAttributes = null;

		if (typeInfo.extensionAttributes != null)
		{
			extensionAttributes = new ArrayList<IEventKDE>();
		}

		if (typeInfo.extensionKDEs != null)
		{
			extensionKDEs = new ArrayList<IEventKDE>();
		}

		OTMappingTypeInformationProperty mappingProp;

		JSONObject jobj = json instanceof JSONObject ? (JSONObject)json : null;
		if (jobj != null)
		{
			for (String jprop : jobj.keySet())
			{
				mappingProp = typeInfo.get(jprop, null);

				if (jobj.has(jprop))
				{
					Object jchild = jobj.get(jprop);
					if (mappingProp != null)
					{
						ReadPropertyMapping(mappingProp, jchild, value, namespaces);
					}
					else if (extensionKDEs != null)
					{
						IEventKDE kde = ReadKDE(jprop, jchild, namespaces);
						extensionKDEs.add(kde);
					}
					else if (extensionAttributes != null)
					{
						IEventKDE kde = ReadKDE(jprop, jchild, namespaces);
						extensionAttributes.add(kde);
					}
				}
			}
		}

		if (typeInfo.extensionAttributes != null)
		{
			typeInfo.extensionAttributes.set(value, extensionAttributes);
		}

		if (typeInfo.extensionKDEs != null)
		{
			typeInfo.extensionKDEs.set(value, extensionKDEs);
		}

		return value;
	}

	private static Object WriteObjectToJToken(Object obj, Boolean isGs1WebVocab)
	{
		if (obj == null)
		{
			return null;
		}
		// check if obj is ArrayList<LanguageString>...
		else if (obj instanceof ArrayList)
		{
			ArrayList l = (ArrayList)obj;
			if (!l.isEmpty())
			{
				JSONArray jArr = new JSONArray();
				for (Object o: l)
				{
					if (o instanceof LanguageString)
					{
						JSONObject j = ((LanguageString)o).toJSON();
						JSONExtensions.put(jArr, j);
					}
					else {
						JSONExtensions.put(jArr, o);
					}
				}
				return jArr;
			}
			else
			{
				return null;
			}
		}
		else if (obj instanceof OffsetDateTime)
		{
			OffsetDateTime dt = (OffsetDateTime)obj;
			return dt.format(DateTimeFormatter.ISO_DATE_TIME);
		}
		else if (obj instanceof UOM)
		{
			UOM uom = (UOM)obj;
			return uom.UNCode;
		}
		else if (obj instanceof Double)
		{
			return obj;
		}
		else if (obj instanceof Boolean)
		{
			return obj;
		}
		else if (obj instanceof Country)
		{
			Country b = (Country)obj;
			return b.abbreviation;
		}
		else if (obj instanceof Duration)
		{
			return StringExtensions.fromDuration((Duration)obj);
		}
		else
		{
			return obj.toString() != null ? obj.toString() : "";
		}
	}

	private static void ReadPropertyMapping(OTMappingTypeInformationProperty mappingProp, Object json, Object value, Map<String, String> namespaces) throws Exception {
		if (mappingProp.isQuantityList)
		{
			IEvent e = (IEvent)value;
			JSONArray jQuantityList = json instanceof JSONArray ? (JSONArray)json : null;
			if (jQuantityList != null)
			{
				for (Object o : jQuantityList)
				{
					if (o instanceof JSONObject)
					{
						JSONObject jQuantity = (JSONObject) o;
						if (!jQuantity.has("epcClass"))
						{
							throw new Exception("Cannot deserialize quantityElement because epcClass not on JSON object.");
						}

						EPC epc = new EPC((jQuantity.getString("epcClass")));

						EventProduct product = new EventProduct(epc);
						product.Type = mappingProp.productType;
						product.Quantity = JSONExtensions.readMeasurement(jQuantity);

						e.addProduct(product);
					}
				}
			}
		}
		else if (mappingProp.isEPCList)
		{
			IEvent e = (IEvent)value;
			JSONArray jEPCList = json instanceof JSONArray ? (JSONArray)json : null;
			if (jEPCList != null)
			{
				for (Object jEPC : jEPCList)
				{
					EPC epc = new EPC(jEPC.toString());
					EventProduct product = new EventProduct(epc);
					product.Type = mappingProp.productType;
					e.addProduct(product);
				}
			}
		}
		else if (mappingProp.isArray)
		{
			List list = (List)ReflectionUtility.constructType(mappingProp.field.getType());

			if (mappingProp.isRepeating && !(json instanceof JSONArray))
			{
				String v = json.toString();
				if (!(v == null || v.isBlank()))
				{
					Object o = ReadObjectFromString(v, mappingProp.itemType);
					list.add(o);
				}
			}
			else
			{
				JSONArray jArr = json instanceof JSONArray ? (JSONArray)json : null;
				if (jArr != null)
				{
					if (mappingProp.itemType == LanguageString.class)
					{
						list = LanguageString.fromJSON(jArr);
					}
					else
					{
						for (Object j : jArr)
						{
							if (mappingProp.isObject)
							{
								Object o = FromJson(j, mappingProp.itemType, namespaces);
								list.add(o);
							}
							else
							{
								Object o = ReadObjectFromString(j.toString(), mappingProp.itemType);
								list.add(o);
							}
						}
					}
				}
			}

			mappingProp.field.set(value, list);
		}
		else if (mappingProp.isObject)
		{
			Object o = FromJson(json, mappingProp.field.getType(), namespaces);
			mappingProp.field.set(value, o);
		}
		else if (ReflectionUtility.isListOf(mappingProp.field.getType(), LanguageString.class))
		{

		}
		else
		{
			String v = json.toString();
			if (!(v == null || v.isBlank()))
			{
				Object o = ReadObjectFromString(v, mappingProp.field.getType());
				mappingProp.field.set(value, o);
			}
		}
	}

	private static Object ReadObjectFromString(String value, Class t) throws Exception {
		return ReflectionUtility.parseFromString(t, value);
	}

	private static IEventKDE ReadKDE(String name, Object value, Map<String, String> namespaces) throws Exception {

		IEventKDE kde = null;

		//if not, check if it is a simple value or an object
		if (kde == null)
		{
			String ns = "";
			if (name.contains(":"))
			{
				ns = StringExtensions.First(name.split(java.util.regex.Pattern.quote(":"), -1));
				name = StringExtensions.Last(name.split(java.util.regex.Pattern.quote(":"), -1));

				if (!namespaces.containsKey(ns))
				{
					throw new RuntimeException("The KDE has a namespace prefix, but there is no such namespace in the dictionary. " + ns);
				}
				ns = namespaces.get(ns);
			}

			if (value instanceof JSONObject || value instanceof JSONArray)
			{
				kde = new EventKDEObject(ns, name);
			}
			//else if simple value, then we will consume it as a string
			else
			{
				kde = new EventKDEString(ns, name);
			}
		}

		if (kde != null)
		{
			kde.setFromJson(value);
		}
		else
		{
			throw new RuntimeException("Failed to initialize KDE from JSON = " + value.toString());
		}

		return kde;
	}
}

File: Java/OpenTraceability/src/main/opentraceability/mappers/opentraceabilityjsonldmapper.kt
package mappers

import com.fasterxml.jackson.core.JsonToken
import interfaces.IEvent
import interfaces.IEventKDE
import models.common.LanguageString
import models.events.EPCISVersion
import models.events.EventProduct
import models.events.kdes.EventKDEObject
import models.events.kdes.EventKDEString
import models.identifiers.EPC
import models.identifiers.GLN
import models.identifiers.GTIN
import models.identifiers.PGLN
import utility.Country
import utility.Measurement
import utility.UOM
import java.util.*
import java.lang.reflect.Type
import java.net.URI
import java.time.OffsetDateTime
import kotlin.reflect.KClass

class OpenTraceabilityJsonLDMapper {
    companion object {

        /**
         * Converts an object into JSON.
         */
        fun ToJson(value: Any?, namespacesReversed: Map<String, String>, required: Boolean = false): JToken? {
            try {
                value?.let {
                    var json: JToken? = JObject()
                    var jpointer: JToken = json

                    val t: Type = value::class.java
                    val typeInfo: OTMappingTypeInformation = OTMappingTypeInformation.GetJsonTypeInfo(t)
                    typeInfo.Properties.filter { it.Version == null || it.Version == EPCISVersion.V2 }.forEach { property ->
                        property.Property.getValue(value)?.let { obj ->
                            var jvaluepointer: JToken = jpointer
                            val xchildname: String = property.Name

                            when {
                                property.IsQuantityList -> {
                                    var products: List<EventProduct> = obj as List<EventProduct>
                                    products = products.filter { it.Quantity != null && it.Type == property.ProductType }
                                    if (products.isNotEmpty()) {
                                        val xQuantityList: JArray = JArray()
                                        for (product in products) {
                                            if (product.EPC != null && product.Quantity != null) {
                                                val xQuantity: JObject = JObject()
                                                xQuantity["epcClass"] = product.EPC.toString()
                                                xQuantity["quantity"] = product.Quantity.Value
                                                if (product.Quantity.UoM.UNCode != "EA") {
                                                    xQuantity["uom"] = product.Quantity.UoM.UNCode
                                                }
                                                xQuantityList.add(xQuantity)
                                            }
                                        }
                                        jvaluepointer[xchildname] = xQuantityList
                                    }
                                }
                                property.IsEPCList -> {
                                    var products: List<EventProduct> = obj as List<EventProduct>
                                    products = products.filter { it.Quantity == null && it.Type == property.ProductType }
                                    if (products.isNotEmpty() || property.Required) {
                                        val xEPCList: JArray = JArray()
                                        for (product in products) {
                                            product.EPC?.let {
                                                xEPCList.add(it.toString())
                                            }
                                        }
                                        jvaluepointer[xchildname] = xEPCList
                                    }
                                }
                                property.IsArray -> {
                                    val list: MutableList<Any> = obj as MutableList<Any>
                                    val xlist: JArray = JArray()

                                    if (list.isNotEmpty() || property.Required) {
                                        if (property.IsRepeating && list.size == 1) {
                                            val jt: JToken? = WriteObjectToJToken(list[0])
                                            if (jt != null) {
                                                jvaluepointer[xchildname] = jt
                                            }
                                        } else {
                                            for (o in list) {
                                                if (property.IsObject) {
                                                    val xchild: JToken? = ToJson(o, namespacesReversed, property.Required)
                                                    if (xchild != null) {
                                                        xlist.add(xchild)
                                                    }
                                                } else {
                                                    val jt: JToken? = WriteObjectToJToken(o)
                                                    if (jt != null) {
                                                        xlist.add(jt)
                                                    }
                                                }
                                            }

                                            jvaluepointer[xchildname] = xlist
                                        }
                                    }
                                }
                                property.IsObject -> {
                                    val xchild: JToken? = ToJson(obj, namespacesReversed, property.Required)
                                    if (xchild != null) {
                                        jvaluepointer[xchildname] = xchild
                                    }
                                }
                                else -> {
                                    val jt: JToken? = WriteObjectToJToken(obj)
                                    if (jt != null) {
                                        jvaluepointer[xchildname] = jt
                                    }
                                }
                            }
                        }
                    }

                    typeInfo.ExtensionKDEs?.getValue(value)?.let { obj ->
                        if (obj is MutableList<IEventKDE>) {
                            val kdes: MutableList<IEventKDE> = obj
                            for (kde in kdes) {
                                kde.GetJson()?.let { xchild ->
                                    var name: String = kde.Name
                                    if (kde.Namespace != null) {
                                        if (!namespacesReversed.containsKey(kde.Namespace)) {
                                            throw Exception("The namespace ${kde.Namespace} is not recognized in the EPCIS Document / EPCIS Query Document.")
                                        }
                                        name = namespacesReversed[kde.Namespace] + ":" + name
                                    }

                                    jpointer[name] = xchild
                                }
                            }
                        }
                    }

                    typeInfo.ExtensionAttributes?.getValue(value)?.let { obj ->
                        if (obj is MutableList<IEventKDE>) {
                            val kdes: MutableList<IEventKDE> = obj
                            for (kde in kdes) {
                                kde.GetJson()?.let { xchild ->
                                    var name: String = kde.Name
                                    if (kde.Namespace != null) {
                                        if (!namespacesReversed.containsKey(kde.Namespace)) {
                                            throw Exception("The namespace ${kde.Namespace} is not recognized in the EPCIS Document / EPCIS Query Document.")
                                        }
                                        name = namespacesReversed[kde.Namespace] + ":" + name
                                    }

                                    jpointer[name] = xchild
                                }
                            }
                        }
                    }

                    return json
                }

                return null
            } catch (ex: Exception) {
                val e = Exception("Failed to parse json. value=$value", ex)
                OTLogger.error(e)
                throw e
            }
        }


        /**
         * Converts a JSON object into the generic type specified.
         */
        inline fun <reified T> FromJson(json: JToken, namespaces: Map<String, String>): T {
            val o: T = FromJson(json, T::class.java, namespaces) as T
            return o
        }


        /**
         * Converts a JSON object into the type specified.
         */
        fun FromJson(json: JToken, type: Type, namespaces: Map<String, String>): Any {
            val value: Any = type.javaClass.kotlin.objectInstance ?: throw Exception("Failed to create instance of type ${type.typeName}")

            try {
                val typeInfo: OTMappingTypeInformation = OTMappingTypeInformation.GetJsonTypeInfo(type)

                var extensionKDEs: MutableList<IEventKDE>? = null
                var extensionAttributes: MutableList<IEventKDE>? = null

                if (typeInfo.ExtensionAttributes != null) {
                    extensionAttributes = mutableListOf()
                }

                if (typeInfo.ExtensionKDEs != null) {
                    extensionKDEs = mutableListOf()
                }

                var mappingProp: OTMappingTypeInformationProperty? = null

                val jobj: JObject? = json as? JObject
                if (jobj != null) {
                    for (jprop in jobj.properties()) {
                        mappingProp = typeInfo[jprop.Name]

                        if (mappingProp != null && mappingProp.Property.setMethod == null) {
                            continue
                        }

                        val jchild: JToken? = jobj[jprop.Name]
                        if (jchild != null) {
                            when {
                                mappingProp != null -> {
                                    ReadPropertyMapping(mappingProp, jchild, value, namespaces)
                                }
                                extensionKDEs != null -> {
                                    val kde: IEventKDE = ReadKDE(jprop.Name, jchild, namespaces)
                                    extensionKDEs.add(kde)
                                }
                                extensionAttributes != null -> {
                                    val kde: IEventKDE = ReadKDE(jprop.Name, jchild, namespaces)
                                    extensionAttributes.add(kde)
                                }
                            }
                        }
                    }
                }

                typeInfo.ExtensionAttributes?.setValue(value, extensionAttributes)
                typeInfo.ExtensionKDEs?.setValue(value, extensionKDEs)
            } catch (ex: Exception) {
                OTLogger.error(ex)
                throw ex
            }

            return value
        }


        fun WriteObjectToJToken(obj: Any?): JToken? {
            return when(obj) {
                null -> null
                is List<LanguageString> -> {
                    val json = obj.toJson()
                    JArray.parse(json)
                }
                is OffsetDateTime -> {
                    obj.toString("O")
                }
                is UOM -> {
                    obj.UNCode
                }
                is Double, is Boolean -> {
                    JToken.fromObject(obj)
                }
                is Country -> {
                    obj.Abbreviation
                }
                is TimeSpan -> {
                    if (obj.ticks < 0)
                        "-${obj.negate().totalHours.toString("#00")}:${obj.minutes.toString("00")}"
                    else
                        "+${obj.TotalHours.toString("#00")}:${obj.minutes.toString("00")}"
                }
                else -> obj.toString() ?: ""
            }
        }

        fun ReadPropertyMapping(mappingProp: OTMappingTypeInformationProperty, json: JToken, value: Any, namespaces: Map<String, String>) {
            when {
                mappingProp.IsQuantityList -> {
                    val e = value as IEvent
                    val jQuantityList = json as? JArray
                    jQuantityList?.forEach {
                        val jQuantity = it as JObject
                        val epc = EPC(jQuantity["epcClass"]?.value<String>() ?: "")

                        val product = EventProduct(epc).apply {
                            Type = mappingProp.ProductType
                            Quantity = Measurement(jQuantity.value<Double>("quantity"), jQuantity.value<String>("uom") ?: "EA")
                        }

                        e.addProduct(product)
                    }
                }
                mappingProp.IsEPCList -> {
                    val e = value as IEvent
                    val jEPCList = json as? JArray
                    jEPCList?.forEach {
                        val epc = EPC(it.toString())
                        val product = EventProduct(epc).apply {
                            Type = mappingProp.ProductType
                        }
                        e.addProduct(product)
                    }
                }
                mappingProp.IsArray -> {
                    var list = mappingProp.Property.getValue(value) as? MutableList<*>
                    if (list == null) {
                        list = (mappingProp.Property.PropertyType.kotlin.objectInstance as? MutableList<*>)
                            ?: throw Exception("Failed to create instance of ${mappingProp.Property.PropertyType.typeName}")

                        mappingProp.Property.setValue(value, list)
                    }

                    val itemType = list::class.java.genericSuperclass as? Class<*>

                    if (mappingProp.IsRepeating && json !is JArray) {
                        val v = json.toString()
                        if (v.isNotBlank()) {
                            val o = ReadObjectFromString(v, itemType)
                            list.add(o)
                        }
                    } else {
                        val jArr = json as? JArray
                        jArr?.forEach {
                            val o = if (mappingProp.IsObject) {
                                FromJson(it, itemType, namespaces)
                            } else {
                                ReadObjectFromString(it.toString(), itemType)
                            }
                            list.add(o)
                        }
                    }
                }
                mappingProp.IsObject -> {
                    val o = FromJson(json, mappingProp.Property.PropertyType, namespaces)
                    mappingProp.Property.setValue(value, o)
                }
                mappingProp.Property.PropertyType == List::class.java -> {
                    val languageStrings: List<LanguageString>? = json.toString().fromJson()
                    if (languageStrings != null) {
                        mappingProp.Property.setValue(value, languageStrings)
                    }
                }
                else -> {
                    val v = json.toString()
                    if (v.isNotBlank()) {
                        val o = ReadObjectFromString(v, mappingProp.Property.PropertyType)
                        mappingProp.Property.setValue(value, o)
                    }
                }
            }
        }



        fun ReadObjectFromString(value: String, t: KClass<*>): Any {
            return try {
                when(t) {
                    OffsetDateTime::class -> {
                        val dt = value.TryConvertToDateTimeOffset() ?: throw Exception("Failed to convert string to datetimeoffset where value = $value")
                        dt
                    }
                    UOM::class -> {
                        val uom = UOM.LookUpFromUNCode(value)
                        uom
                    }
                    Boolean::class -> {
                        val v = value.toBoolean()
                        v
                    }
                    Double::class -> {
                        val v = value.toDouble()
                        v
                    }
                    URI::class -> {
                        val v = URI(value)
                        v
                    }
                    TimeSpan::class -> {
                        val ts = if (value.startsWith("+")) TimeSpan.parse(value.drop(1)) else TimeSpan.parse(value)
                        ts
                    }
                    EventAction::class -> {
                        val action = EventAction.valueOf(value)
                        action
                    }
                    PGLN::class -> {
                        val pgln = PGLN(value)
                        pgln
                    }
                    GLN::class -> {
                        val gln = GLN(value)
                        gln
                    }
                    GTIN::class -> {
                        val gtin = GTIN(value)
                        gtin
                    }
                    EPC::class -> {
                        val epc = EPC(value)
                        epc
                    }
                    Country::class -> {
                        val c = Countries.parse(value)
                        c
                    }
                    else -> value
                }
            } catch (ex: Exception) {
                val e = Exception("Failed to convert string into object. value=$value and t=$t", ex)
                OTLogger.error(e)
                throw e
            }
        }

        fun ReadKDE(name: String, json: JToken, namespaces: Map<String, String>): IEventKDE {
            var kde: IEventKDE? = null
            var ns = ""
            var realName = name

            if (name.contains(":")) {
                val split = name.split(":")
                ns = split.first()
                realName = split.last()

                if (!namespaces.containsKey(ns)) {
                    throw Exception("The KDE has a namespace prefix, but there is no such namespace in the dictionary. $ns")
                }
                ns = namespaces[ns] ?: ""
            }

            if (json is JObject || json is JArray) {
                kde = EventKDEObject(ns, realName)
            } else {
                kde = EventKDEString(ns, realName)
            }

            kde?.setFromJson(json) ?: throw Exception("Failed to initialize KDE from JSON = ${json.toString()}")

            return kde
        }


    }
}
